proxy/src/timeout.rs

```diff
@@ -1,38 +1,44 @@
 // #![deny(missing_docs)]
-use futures::{Async, Future, Poll};
+use futures::{Future, Poll};
 
 use std::error::Error;
 use std::{fmt, io};
-use std::time::Duration;
+use std::time::{Duration, Instant};
+use std::sync::Arc;
 
 use tokio_connect::Connect;
-use tokio_core::reactor::{Timeout as ReactorTimeout, Handle};
-use tokio_io;
+use tokio::timer::{self, Deadline, DeadlineError};
+use tokio::io::{AsyncRead, AsyncWrite};
 use tower_service::Service;
 
+pub type Name = Option<Arc<str>>;
+
 /// A timeout that wraps an underlying operation.
 #[derive(Debug, Clone)]
-pub struct Timeout<U> {
-    inner: U,
+pub struct Timeout<T> {
+    inner: T,
     duration: Duration,
-    handle: Handle,
+    name: Name,
```

Comment:
This was added to aid with debugging a bug while working on this branch. I found it valuable, so I decided to leave it in; hopefully that was acceptable.

Reply:
Oh interesting! Have an example of it's use that I can look at?

Reply:
Sure! Just as a quick example, the expected timeouts in some of the discovery tests now log
```
ERROR 2018-05-08T23:52:35Z: conduit_proxy: turning binding outbound client timed out after 100ms into 500
```

They used to say
```
ERROR 2018-05-08T23:52:35Z: conduit_proxy: turning operation timed out after 100ms into 500
```
regardless of what the operation that timed out was.

I found this really useful while debugging an issue which caused the timeouts to fail, and I wasn't sure where the failing timeouts were being created. 

I suspect that if we have issues in the future relating to timeouts (whether due to the timeout failing or due to things timing out), having the most visible log entry associated with the error actually tell us what errored may save us some work debugging.

Reply:
I'm sure it's useful! I should have clarified, I meant the actual code invocation. I'm just wondering if there's a need to allocate a string for every single timeout, or if a `&'static str` could be enough.

Reply:
Changed in 644df5a387edb7b3b45506bdc2a714cb9eed8f88 :)